### PR TITLE
EVG-6396: rework authentication

### PR DESCRIPTION
**Pull request by dominoweir**
This set of changes includes modifying spruce based on my changes to the evergreen.js library, adding logout functionality, eliminating the need for an api key and username in a config file, and some merging from master in to this branch.

### CODE REVIEW

File: src/components/app/App.tsx [hunk 1]
```diff
@@ -13,6 +13,7 @@ import { PatchContainer } from '../patch/PatchContainer';
 interface State {
   APIClient: rest.Evergreen;
   MenuAnchor?: HTMLElement;
+  username?: string;
```
[the remark this answers is not in the document]
Comment (dominoweir) [reply src/components/app/App.tsx]: Hmm... I'm not sure. The user already knows what their username is, so it's not like they need to be reminded of it. Maybe we can incorporate it the same way the current Evergreen UI does, where it's shown as part of the "here's what your yaml file should contain" box on the user settings page.

File: src/components/login/Login.tsx [hunk 2]
```diff
@@ -56,12 +61,17 @@ export class Login extends React.Component<Props, State> {
...
               id="password"
               label="password"
-              type="string"
+              type="password"
               fullWidth={true}
               required={true}
               onChange={this.updateSingleField(this.state, "password")}
               value={this.state.password}
             />
+            <div hidden={this.state.hideErrorMessage}>
```
[the remark this answers is not in the document]
Comment (dominoweir) [reply src/components/login/Login.tsx]: Good idea, I removed that hideErrorMessage boolean in the state and replaced it with an errorText string.

File: src/components/login/Login.tsx [hunk 3]
```diff
@@ -77,40 +87,56 @@ export class Login extends React.Component<Props, State> {
   }
 
   private handleClickOpen = () => {
-    this.setState({
-      open : true,
-    });
+    if (this.state.buttonText === "Log In") {
```
[the remark this answers is not in the document]
Comment (dominoweir) [reply src/components/login/Login.tsx]: Fixed

File: src/components/app/App.tsx [hunk 4]
```diff
@@ -31,45 +32,51 @@ export class Evergreen extends React.Component<Props, State> {
...
-            </Toolbar>
-          </AppBar>
-          <div className="App-intro">
-            <Route path="/admin" render={admin} />
-            <Route path="/config" render={config} />
-            <Route path="/patches" render={patches} />
-          </div>
-        </HashRouter>
+        <UserContextConsumer>
+          {props => {
```
Comment (john-m-liu) [inline src/components/app/App.tsx]: nit: this isn't used so it can just be ()

**Pull request merged**